src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentStepImpl.java

```diff
@@ -0,0 +1,220 @@
+ *
+ * @since deployments
+ */
+public class DeploymentStepImpl extends Step implements DeploymentStep {
+
+    private final String environmentName;
+
+    private String environmentKey;
+    private BitbucketDeploymentEnvironmentType environmentType;
+    private String environmentUrl;
+
+    @DataBoundConstructor
+    public DeploymentStepImpl(@CheckForNull String environmentName) {
+        this.environmentName = stripToNull(environmentName);
+    }
+
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        Run<?, ?> run = context.get(Run.class);
+        if (run == null) {
+            throw new IllegalArgumentException("StepContext does not contain the Run");
+        }
+        TaskListener taskListener = context.get(TaskListener.class);
+        if (taskListener == null) {
+            throw new IllegalArgumentException("StepContext does not contain the TaskListener");
+        }
+        BitbucketDeploymentEnvironment environment = getEnvironment(run, taskListener);
+        return new DeploymentStepExecution(environment, context);
+    }
+
+    @Override
+    public DescriptorImpl getDescriptor() {
+        return (DescriptorImpl) super.getDescriptor();
+    }
+
+    @Override
+    public BitbucketDeploymentEnvironment getEnvironment(Run<?, ?> run, TaskListener listener) {
+        return DeploymentStepUtils.getEnvironment(this, run, listener);
+    }
+
+    @Override
+    public String getEnvironmentKey() {
+        return environmentKey;
+    }
+
+    @DataBoundSetter
+    public void setEnvironmentKey(@CheckForNull String environmentKey) {
+        this.environmentKey = getOrGenerateEnvironmentKey(environmentKey);
+    }
+
+    @CheckForNull
+    @Override
+    public String getEnvironmentName() {
+        return environmentName;
+    }
+
+    @CheckForNull
+    @Override
+    public String getEnvironmentType() {
+        return environmentType == null ? null : environmentType.name();
+    }
+
+    @DataBoundSetter
+    public void setEnvironmentType(@CheckForNull String environmentType) {
+        this.environmentType = normalizeEnvironmentType(environmentType);
+    }
+
+    @CheckForNull
+    @Override
+    public String getEnvironmentUrl() {
+        return environmentUrl;
+    }
+
+    @DataBoundSetter
+    public void setEnvironmentUrl(@CheckForNull String environmentUrl) {
+        this.environmentUrl = stripToNull(environmentUrl);
+    }
+
+    @Extension
+    public static class DescriptorImpl extends StepDescriptor {
+
+        @VisibleForTesting
+        static final FormValidation FORM_VALIDATION_OK = FormValidation.ok();
+
+        @Inject
+        private BitbucketDeploymentFactory bitbucketDeploymentFactory;
+        @Inject
+        private DeploymentPoster deploymentPoster;
+        @Inject
+        private JenkinsProvider jenkinsProvider;
+
+        @POST
+        public FormValidation doCheckEnvironmentName(@AncestorInPath @CheckForNull Item context,
```

Comment:
I'm not too sure how it works in the UI and I know we can auto generate the environment key, but do we also need to check permissions for the environment key?

Reply:
Do you mean Jenkins would prevent a user from defining a deployment notifier step with "Wilson" as the environment key if the user does not have permission to deploy to the "Wilson" environment (but may have permission to deploy to other environments)? AFAIK, environment keys do not have access control lists in Jenkins nor in Bitbucket Data Center. Even if such ACLs were added to Bitbucket DC, I think it would be OK to let the user define the step in Jenkins and just let it fail at run time if the credential configured for the Jenkins job does not have access.

Reply:
The UI calls this method as part of form validation so that it can display an error if the value doesn't pass validation.
However, we are essentially exposing a REST endpoint that anyone can hit, so it is good practice to add some basic permission checks on the endpoint so that 'not just anyone' can access it. In this case the endpoint doesn't actually do anything expensive, or expose anything confidential, so the permission checks aren't super important. But they're good to have regardless.

You raise a valid point about doing form validation on the environment key. I didn't add any validation because we can generate it if it is not provided. However, we _could_ do some extra validation to make sure it doesn't contain any unexpected characters, and ensure that it is not too long. However, we're not doing any of those checks on the other fields either.

Reply:
Thanks for the explanation everyone 😃